### PR TITLE
Drop support for IID_ICorDebugProcess10 and fix thread suspend logic

### PR DESCRIPTION
**Pull request by sdmaclea**
See https://github.com/dotnet/runtime/pull/44539#discussion_r521580819

### CODE REVIEW

File: src/coreclr/src/vm/threadsuspend.cpp [hunk 1]
```diff
@@ -5943,16 +5943,7 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
...
-            // When the debugger is synchronizing, trying to perform a GC could deadlock. The GC has the
-            // threadstore lock and synchronization cannot complete until the debugger can get the
-            // threadstore lock. However the GC can not complete until it sends the BeforeGarbageCollection
-            // event, and the event can not be sent until the debugger is synchronized. In order to break
-            // this deadlock cycle the GC must give up the threadstore lock, allow the debugger to synchronize,
-            // then try again.
-            // 
-            || (CORDebuggerAttached() &&
-                (g_pDebugInterface->ThreadsAtUnsafePlaces() || g_pDebugInterface->IsSynchronizing()))
+            || (CORDebuggerAttached() && g_pDebugInterface->ThreadsAtUnsafePlaces())
```
Comment (sdmaclea) [inline src/coreclr/src/vm/threadsuspend.cpp:5946]: This is a simplification of @kouvel's change as proposed in  https://github.com/dotnet/runtime/pull/44539#discussion_r521580819

**Review by noahfalk — APPROVED** [review]
LGTM : )

**Comment by hoyosjs** [comment]
@chuckries So going forward:

New versions of a debugger will just use the new mechanism for 5.0.1+ and notifications for 3.1 and below, where there is no risk of deadlocks?

On older versions 6.0+ will fail the QI so they just can't use DBP, 5.0.0 will potentially deadlock, 5.0.1 should probably also remove the QI to make it possible to just have the behavior in a good state for 5.0.x, and 3.1 work as usual.

Is this correct?

If so, this PR looks good to me

**Comment by sdmaclea** [comment]
@hoyosjs Yes the plan is to back port this and #44549 to 5.0.  I didn't include the more general clean up to remove GC Event notifications in this PR to make it easier to get through servicing.

**Review by chuckries — APPROVED** [review]

**Review by hoyosjs — APPROVED** [review]

**Comment by kouvel** [comment]
If an unpatched VS tries to add a data breakpoint to a patched runtime, would it fail?

**Comment by kouvel** [comment]
Oh I see it would look like the patched runtime does not support DBs, think I got it now. LGTM too, thanks!

**Pull request merged**